Add drag & drop installation

asika32764 opened the pull request:
### Summary of Changes

Add drag & drop file to install extension package.

The drag zone is whole page of the **Extensions -> Manage ->  Install**  view and it ~~works for 4 tabs~~ now only works for `Upload package file` tab.

### Testing Instructions

Go to **Extensions -> Manage ->  Install** and drag file from desktop to browser. Joomla will auto upload it to install.

### Images

![drag-upload](https://cloud.githubusercontent.com/assets/1639206/24330199/e906b01e-124b-11e7-8039-720dcaccb774.gif)

asika32764 committed:
Add drag & drop installation

asika32764 committed:
Fix drag issue

asika32764 committed:
Add more error message

asika32764 committed:
Typo

dgt41 commented:
Formdata is not supported in IE8 http://caniuse.com/#search=formdata
so we need some sort of fallback (in tinymce we have a message)

brianteeman commented:
First of all thank you very much for doing this. I drag and drop all the time to upload extensions and its a pain to hit the very small target so this will make a great difference

We have drag and drop in one other place in Joomla which is for images within tinmyce and it would be good if you could make the language string here match. 

So can you change the string to "Drag and Drop ...." not just Drag

Secondly I can see in the tinymce plugin for dragndrop  (media\editors\tinymce\plugins\jdragdrop\plugin.js) that we have a check and an error if the users browser doesnt support drag drop. So I presume we need a similar to check here?

asika32764 committed:
CS

brianteeman commented:
I see that @dgt41 was writing the same thing at the same time as me ;)

asika32764 commented:
Thanks for the feedback, I'll try fix the IE8 issues later.

asika32764 committed:
Style fix

asika32764 committed:
Fix from review

asika32764 commented:
Added a hint to notice IE8 users which is not support drag&drop upload. I don't have Windows now, could anyone help me to test?

The language key wrote by @dg41 is only for Tinymce plugin and hard coded the `image` words in language string, so I added 2 new language keys for this feature in com_installer.

asika32764 committed:
Add return

infograf768 commented:
It works, but one has no idea that Drag and Drop is possible at looking at the UI.
I suggest adding this information on the page.

brianteeman commented:
@infograf768 the same could be said of the drag and drop of images in
tinmyce ;) I dont deel it is necessary as it is a standard web pattern now

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On 26 March 2017 at 11:25, infograf768 <notifications@github.com> wrote:

> It works, but one has no idea that Drag and Drop is possible at looking at
> the UI.
> I suggest adding this information on the page.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/14905#issuecomment-289271074>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8RwfLpGmVHvD8Kii8vaRHNWhLaClks5rpj0HgaJpZM4MpX-_>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
https://brian.teeman.net/ <http://brian.teeman.net/>

brianteeman commented on administrator/language/en-GB/en-GB.com_installer.ini:
```diff
@@ -12,6 +12,8 @@ COM_INSTALLER_CONFIRM_UNINSTALL="Are you sure you want to uninstall? Confirming
 COM_INSTALLER_CURRENT_VERSION="Installed"
 COM_INSTALLER_DISCOVER_FILTER_SEARCH_DESC="Search in discovered extension name. Prefix with ID: to search for an extension ID."
 COM_INSTALLER_DISCOVER_FILTER_SEARCH_LABEL="Search Discovered Extensions"
+COM_INSTALLER_DRAG_ERR_UNSUPPORTEDBROWSER="Drag and drop file upload is not available for your browser. Please consider using a fully HTML5 compatible browser"
```
perfect text - please add a . at the end

asika32764 replied:
Also fixed the original text from TinyMCE

asika32764 commented:
> It works, but one has no idea that Drag and Drop is possible at looking at the UI.
I suggest adding this information on the page.

I have considered about this but I didn't implement this message yet, I'm open to let people decide what to do.

franz-wohlkoenig commented:
@infograf768 i guess most Users are surprised that its not working (like this PR allow). So you have experienced Users and they should got Info by Release-Data.

asika32764 committed:
Add `.`

infograf768 commented:
When we edit TinyMCE plugin, it is indicated.
![screen shot 2017-03-26 at 12 29 09](https://cloud.githubusercontent.com/assets/869724/24330514/f886ae12-121f-11e7-9565-80af1b6a6bfb.png)

It is indeed not indicated when using tinyMCE as an editor to drop images into the textarea.

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 96417925a117b03e6d4ad0e75981cec5b431c5c2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14905">issues.joomla.org/tracker/joomla-cms/14905</a>.</sub>

brianteeman commented:
@infograf768 that is for the drag and drop of moving the tinymce buttons. NOT for the drag and drop image uploading ;)

brianteeman commented:
@franz-wohlkoenig it has always worked (a bit) - just that you had to drop it on the button and then click on the upload button. This PR makes it work as it should

brianteeman commented:
would something like this be acceptable to you

<img width="557" alt="screenshotr11-39-30" src="https://cloud.githubusercontent.com/assets/1296369/24330578/ea00ddec-1218-11e7-958b-c612cdd9947b.png">

jeckodevelopment commented:
In many applications, when there's the opportunity to drag and drop, there's a kind of dashed border to highlight the area in which you can drag and drop the file/whatever

Bakual commented:
> The drag zone is whole page of the Extensions -> Manage -> Install view and it works for 4 tabs.

Imho it should only work on the "Upload Package File" tab. To me it doesn't make much sense and could beconfusing to be able to drop a file into the "Install by URL", "Install from Folder" or "Install from Web" tab.
Also, the code should be added by the plugin responsible for that tab (https://github.com/joomla/joomla-cms/blob/staging/plugins/installer/packageinstaller/tmpl/default.php) and not by com_installer.

dgt41 commented:
@asika32764 can you move the code to the file form field (maybe with an option allowdnd) ?
This way we make it available everywhere, not only in this page

brianteeman commented:
oh i didnt spot it worked in all tabs - that is indeed confusing

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On 26 March 2017 at 12:08, Thomas Hunziker <notifications@github.com> wrote:

> The drag zone is whole page of the Extensions -> Manage -> Install view
> and it works for 4 tabs.
>
> Imho it should only work on the "Upload Package File" tab. To me it
> doesn't make much sense and could beconfusing to be able to drop a file
> into the "Install by URL", "Install from Folder" or "Install from Web" tab.
> Also, the code should be added by the plugin responsible for that tab (
> https://github.com/joomla/joomla-cms/blob/staging/plugins/installer/
> packageinstaller/tmpl/default.php) and not by com_installer.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/14905#issuecomment-289273098>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8dNQ1k9NDzJvmnKJibv3a1XXhSSkks5rpkcbgaJpZM4MpX-_>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
https://brian.teeman.net/ <http://brian.teeman.net/>

brianteeman commented:
@jeckodevelopment look at how its done in the tinymce plugin we already have for image uploads. You only get the box when you are dragging and dropping - not before

Bakual commented:
> an you move the code to the file form field (maybe with an option allowdnd) ?
This way we make it available everywhere, not only in this page

It's an idea, but it would have to work also when multiple file fields are present in a form. You can't have the whole form be the trigger area in this case. You probably need an attribute which specifies the area identifier.

dgt41 commented:
@Bakual I guess we ~can~ append (programmatically) some unique classes to the drop area and have the script loop through the classes and do it's magic

asika32764 commented:
> Imho it should only work on the "Upload Package File" tab.

I didn't put code in plugin because I set whole page as the drag zone. And the `upload from file` function is hard coded in com_installer `InstallerModelInstall`, then the ajax upload must add a new task to controller, it is deep coupled to core installer component, so I decided to make the feature works for all tabs. This concept also inspired by phpMyAdmin, it can drag file in any pages to import SQL file.

I can move code to plugin and make the drop zone only in `install from upload` tab content, but we need a box with dashed styling, I'm not sure how should I do this layout (what size? color? and etc...).


> can you move the code to the file form field (maybe with an option allowdnd) ?
This way we make it available everywhere, not only in this page

@dgt41 It is a good idea for a drag & drop file field, but it is too big to me to implement this feature since I just got one night free to do this. Let's make this PR work first and leave the new field for future PRs.

asika32764 commented:
@dgt41 

About the drag & drop field, since Javascript is not able to inject file path to HTML file input, we have only two way that use ajax to send POST uploading or convert file to base64 and store in text field.

I have implemented this feature in my RAD framework, see:
![default](https://cloud.githubusercontent.com/assets/1639206/24330954/ab3f5cfc-125c-11e7-9df4-21d4b9aa1f9f.jpg)

But we must consider about this is not a standard way for JForm process, developer must manually handle the ajax or base64 string for every fields.

asika32764 committed:
Fix based on review

asika32764 commented:
**UPDATE**

Based on discussion, moved code to `plg_installer_packageinstaller` and limited this function only work for `Upload Install` tab, also added the message to notice user.

I still set whole page as drag zone, whether we should create a dashed box depends on future discussion.

asika32764 committed:
Add return url

asika32764 committed:
Convert loading cover operation into object

asika32764 committed:
MISC fixes

franz-wohlkoenig commented:
@asika32764 lets chose the Way People expect: i guess its a dashed box, if this is best Practice.

infograf768 commented on administrator/components/com_installer/views/install/tmpl/default.php:
```diff
@@ -43,14 +43,26 @@
-		.css("display", "none")
-		.css("margin-top", "-10px");
+		JoomlaInstaller.getLoadingOverlay()
+            .css("top", outerDiv.position().top - $(window).scrollTop())
```
Could you use tabs instead of spaces to indent these js and css?

asika32764 replied:
Fix in https://github.com/joomla/joomla-cms/pull/14924

asika32764 commented:
Create a new PR in https://github.com/joomla/joomla-cms/pull/14924 that we can discuss for the new box layout.

infograf768 commented:
Closed as we have new PR https://github.com/joomla/joomla-cms/pull/14924<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14905">issues.joomla.org/tracker/joomla-cms/14905</a>.</sub>
